sdk/trace/simple_span_processor.go

```diff
@@ -90,6 +90,16 @@ func (ssp *simpleSpanProcessor) Shutdown(ctx context.Context) error {
 
 		select {
 		case err = <-done:
+			// It is possible for the exporter to have immediately shut down and
+			// the context to already be done. In that case this select statement
+			// will randomly choose a case. This could result in a different
+			// response for similar scenarios. Instead, double check if the
+			// context is done here and return that error if the exporter shut
+			// down did not return one.
+			if err == nil {
+				// If ctx is not done yet, ctx.Err returns nil.
+				err = ctx.Err()
+			}
```

Comment:
Exiting before we pass the shut down signal to the enclosed exporter seems like it could indeed cause a leak. For that reason I added this here instead of at the start.

Reply:
you can consider using `github.com/uber-go/goleak` in the test to avoid a regression bug

feel free to resolve the comment

Reply:
> maybe we should wait for `<-done:` first (without `select`) to avoid a potential leak and then check `ctx.Err();` thinking

This behavior would make us non-compliant with [the spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#shutdown):

> `Shutdown` SHOULD complete or abort within some timeout.

---

> Also, I started to think that it may be actually good that nil was returned if the exporter to have immediately shut down (the shutdown was not canceled, right?)

I can see it both ways, but I think you are right here. The most common path with be reporting these back to the `TracerProvider` and there's no need to report this there. I'll refactor.